### PR TITLE
[FEATURE] Pivot implementation

### LINKED ISSUE #1201: pivot function implementation

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1201 
- [x] tests added and passing

That implementation uses `unstask` function that falls default to pandas in current master, but after #1649 will be merged it will not

### CODE REVIEW

**Comment by dchigarev** [comment]
At `pivot` implementation in query compiler, I rely on that there is a `unstack` function. But that method will be added only when #1649 will be merged. So I just copied it implementation from #1649. Please do not merge this PR until #1649 will be merged and that draft implementation will be removed from my PR.

**Comment by devin-petersohn** [comment]
Ok @dchigarev I will place a blocked tag on this PR. Feel free to remove it once #1649 is merged.

File: modin/backends/pandas/query_compiler.py [hunk 1]
```diff
@@ -1466,6 +1466,55 @@ def compute_groupby(df):
...
+            return df.apply(lambda df: df.set_index(columns).T).droplevel(1)
+
+        # at the reduce phase we will get df with NaN values placed like this:
+        #        A    B    C 
+        # one    1  nan    3
+        # two    4    5  nan    so we want to apply `bfill` fillna method
+        # one  nan    2  nan    fo each group to get desired result
+        # two  nan  nan    6
+        def reduce_func(df):
+            return df.apply(lambda df: df.fillna(method="bfill").iloc[0])
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1505]: Nice, very clean.

**Comment by dchigarev** [comment]
@devin-petersohn I've add some fixes into `pivot` implementation, I haven't tested it for multiple `values`, but for single one it works well. 

But there is some mismatch from pandas in this implementation. If our specified `index` contains `NaN` values, then we would lose that as a group into groupby (pandas does not, because its `pivot` implementation uses different from groupby logic to group insides and columns). For now I've made a dirty fix `index.fillna("NaN")` just to show that it would properly work in that case, but I suppose that it should be mention in design documentation how we would like to handle NaNs if we want to use groupby in our implementation.

**Comment by devin-petersohn** [comment]
@dchigarev pandas 1.1 (August 1) will have a new feature added in pandas-dev/pandas#30584 to keep or drop `nan` vaues from the index. Maybe it is not worth spending too long on a workaround.

**Comment by dchigarev** [comment]
@devin-petersohn I've finished implementing `pivot`. Here is a few notes about implementation:

1. For now handling `nan` groups been doing via `index.fillna("NaN")` just to output some result in that case
2. There is a bug in pandas groupby [#34455](https://github.com/pandas-dev/pandas/issues/34455) related with applying function to groupby object when indices of source dataframe are floats. Since we are using groupby, modin implementation also have that issue, I've solved it by converting `float` indices to `str` indices before groupby and backward after, tests seem to be passed for that case. So that workaround also should be removed after [#34455](https://github.com/pandas-dev/pandas/issues/34455) will be fixed

**Comment by dchigarev** [comment]
@devin-petersohn Yeah, the problem seems to appear with increasing amount of `values` columns and decreasing amount of unique values in `index`.

Reduce part in `GroupbyReduce` is a bottleneck here, I've modified [`groupby_reduce`](https://github.com/modin-project/modin/blob/master/modin/engines/ray/pandas_on_ray/frame/partition_manager.py#L90-L116) this way to do measurements:

(please note if this method of measurement is incorrect)
<details><summary>groupby_reduce</summary>



```python
    @classmethod
    def groupby_reduce(
        cls, axis, partitions, by, map_func, reduce_func
    ):  # pragma: no cover
        from timeit import default_timer as timer
        t1 = timer()

        map_func = ray.put(map_func)
        by_parts = np.squeeze(by)
        if len(by_parts.shape) == 0:
            by_parts = np.array([by_parts.item()])

        print("GRP_preparations:", timer() - t1)
        t1 = timer()

        new_partitions = np.array(
            [
                [
                    PandasOnRayFramePartition(
                        func.remote(
                            part.oid,
                            by_parts[col_idx].oid if axis else by_parts[row_idx].oid,
                            map_func,
                            part.call_queue,
                            by_parts[col_idx].call_queue
                            if axis
                            else by_parts[row_idx].call_queue,
                        )
                    )
                    for col_idx, part in enumerate(partitions[row_idx])
                ]
                for row_idx in range(len(partitions))
            ]
        )

        [y.get() for x in new_partitions for y in x]
        print("GRP_map:", timer() - t1)
        
        t1 = timer()
        res = cls.map_axis_partitions(axis, new_partitions, reduce_func)
        [y.get() for x in res for y in x]
        print("GRP_reduce:", timer() - t1)
        return res
```


</details>

And I've got this result for `(30000, 3)` data set where there is no repeated values in `index`
```
GRP_preparations: 0.0021804869174957275
GRP_map: 0.27529858611524105
GRP_reduce: 13.666183208115399
```
And if you'll look at CPU usage at the reduce part, probably only one core would be used.

Let's look at two corner cases:
1. **We have `index` with one unique value**
image
At this picture we can see a problem of approach "groupby + transpose". If we have source dataframe with long columns and small rows (selected data on which we would do groupby reduce), we would probably get only rows partitions. So applying transpose to every partition in a map phase will generate a lot of `1xN` partitions. But as we know, reduce part is a `map_axis_partitions(0, ...)`, so technically we will apply our reduce function to the whole dataframe without any scaling.

Let's say we have `(30000, 3)` DataFrame with a single unique value in `index` and 60 cores.

- At stage 1 we will have `(30000, 3)` DataFrame with 60 partitions each with `(500, 3)` shape. 
- Applying map function will give us at stage 2 `(60, 30000)` DataFrame with 60 partitions each with `(1, 500)` shape. 
- And at stage 3 we will get `(60, 30000)` DataFrame as an input of our reduce function.

Adding this line before reduce part to trigger resplitting at rows axis, gives increase of performance in 5x times
```python
new_partitions = cls.map_axis_partitions(1, new_partitions, lambda df: df)
```
Execution time for `(30000, 3)` dropped from ~20s to ~4s (which is still slower than pandas). So after map phase, we want to be able to do a resplitting our modin frame to get scaling in the reduce part, but I wonder if `modin_frame` provides such API?

2. **Now let's say that all values in `index` are unique**
Then,
- At stage 1 we will have `(30000x3)` DataFrame with 60 partitions each with `(500, 3)` shape
- Applying map function will give us at stage 2 `(30000, 1)` DataFrame with 60 partitions each with `(500, 1)` shape.  
- And at stage 3 we will get `(30000, 1)` DataFrame as an input of our reduce function.

It seems that groupby + apply for high dimension data is pretty hard for pandas, for example for `(30000, 1)` dataframe with all unique values in `index` that line of code takes about 10s
```python
by = pd_df["_index"]
pd_df.drop(columns=["_index", "_columns"]).groupby(by).apply(lambda df: df)
```

So I think unsplitted data passed to the reduce function is a problem here.

P.S. Also tested the previous implementation of `pivot` via `unstack` (766cc14) and for `(30000, 3)` DataFrame I've got these times:
```
(30000, 3) DataFrame, 10 groups of unique values in index
pandas: 0.12308201845735312
modin (a1c6c90): 16.28071112651378
modin (a1c6c90) + splitting hack: 6.3267082357779145
modin (766cc14): 0.9199626314640045
```
766cc14 still has places to optimize and requires refactoring, this uses draft `unstack` implementation from #1649 
@devin-petersohn may we think of using old implementation?

**Comment by dchigarev** [comment]
@devin-petersohn I've refactored `pivot` implementation via `unstack`, please take a look. Also keep in mind, that this branch uses draft `unstack` implementation so the performance results could be not relevant with the final version (but it's already better than implementation with groupby + reduce).

**Comment by dchigarev** [comment]
@devin-petersohn since `unstack` is not ready yet, can we merge `pivot` as is (with `unstack` defaulting to pandas) and just open an issue to test/improve performance of `pivot` with `unstack` when it will be merged? Same question about `pivot_table` (#1669)

**Comment by dchigarev** [comment]
@devin-petersohn since `unstack` is merged, this PR is not blocked anymore and ready for review.

**Pull request merged**